### ISSUE TITLE
symbol lookup error on target library 

### ISSUE THREAD

**Issue by glen-fine** [rust-lang/cargo#6942]
**Problem**

Getting a symbol lookup error on a library built by RUST 

(pcfw is our application. libwarmer_firmware.so is our library) 

**pcfw: symbol lookup error: /usr/lib/libwarmer_firmware.so: undefined symbol: _Z
N7futures9task_impl4core3GET17h819f28b4899bfef6E**

**Steps**

1. Compile using ["-C", "prefer-dynamic"] in .cargo/config [build]
2. Produce a .so that contains some futures code. 
3. run the executable that uses the library on the target.

rustc and cargo only reside on the dev machine. 

rustup --version
rustup 1.18.2 (a0bf3c9cb 2019-05-02)

rustc --version
rustc 1.34.1 (fc50f328b 2019-04-24)

contents of .cargo/config
[target.armv7-unknown-linux-gnueabihf]
ar = "arm-dey-linux-gnueabi-gcc-ar"
linker = "gcc-sysroot"

[build]
rustflags = ["-C", "prefer-dynamic"]

The library compiles without an error. Just some warnings for unused variables.  When I put the resulting libwarmer_firmware.so on the target I get the error I posted above when pcfw executes.  pcfw (also a rust program) was built at the same time but in a different part of our application tree.

I did a 'cargo update' and still get the same error on the target. 

**Possible Solution(s)**

There seems to be a dependency tree that is open ended.

**Notes**
Output of `cargo version`:
cargo --version
cargo 1.34.0 (6789d8a0a 2019-04-01)

The host OS (dev) is Ubuntu 16.04 LTS. 
The target is Linux (running 3.14.79) on a Digi i.MX6Q board. 

The target toolchain is .rustup/toolchains/stable-armv7-unknown-linux-gnueabihf/

I am fairly new to RUST so if this is a newbie error, please let me know.  I scanned this forum for similar problems and found none matching this type of error.

**Comment by glen-fine** [comment]
I did some deeper digging, and found some additional libraries in target/armv7-unknown-linux-gnueabihf/release/deps and copied these .so files over to the target. 

Then the errors changed to 

**/usr/bin/pcfw: symbol lookup error: /usr/bin/pcfw: undefined symbol: _ZN110_$LT
$tokio..runtime..threadpool..task_executor..TaskExecutor$u20$as$u20$tokio_execu
tor..executor..Executor$GT$5spawn17ha09c84a48cd17f4aE
/usr/bin/fwx: symbol lookup error: /usr/bin/fwx: undefined symbol: _ZN53_$L
T$service..WsRemote$u20$as$u20$service..Remote$GT$8incoming17h53e365d5dd5a8507E**

fwx is our message server.  pcfw just interrogates personality boards in our system. 

I also noticed several files in the 'build' subdir. 
**arrayvec-d4bd799cb11c7637       
parking_lot_core-af3b530b214e6126
backtrace-2a7c52c98e9272b6      
rand-1ce2bd854c45345a
backtrace-sys-f34b2b329342ce74  
rand_chacha-ef676e5de319f27b
byteorder-65e2800192b19b85      
rand_pcg-c616f3fd64cc0d82
libc-066000db92c091af           
ryu-9a726284a8304b5a
nix-807f46fb769f7c6d            
serde-cd4de373d1d0f7cb**

Do these need to be copied over too? 
or is there a RUST compile switch I can use to avoid these dependency loops. 

BTW, just for info. my cargo compile is 

**cargo build --target=armv7-unknown-linux-gnueabihf --release**

**Comment by sfackler** [comment]
prefer-dynamic asks the compiler to dynamically link all Rust dependencies, which means you'll need to distribute all of those shared libraries along with your application.

**Comment by glen-fine** [comment]
Thanks Steven.  

However, it almost seems the second I move libraries to the target more dependencies (symbol lookup errors) crop up.  I have copied all the files in rustup/toolchain/armv7-unknown-linux-gnueabihf, the .so files in /deps and I have no idea if and where to copy the ones in /build (or even if they should be copied).  

It is almost as if each library I copy is dependent on, yet again, another library.  I was hoping for a simple resolution.  I have tried removing 'prefer-dynamic' and still these symbol problems still occur, and I make doubly sure the compile is --release.

**Comment by glen-fine** [comment]
Closing as it is the level of dependencies, and what needs to be copied to the target, is what created the problem. 

This might be a good idea to explain in gory detail in some RUST document when using "prefer-dynamic" that a bunch of libraries need to be copied to the target.  Usually under a 'noob' should watch out for ......  

What I found was not only are the executables needed on a target, but so are all the libraries that show up in the release (or debug) subdirs like /deps, and one must also be aware of the libraries in the .rustup/toolchain/<target>  which need to be copied to the target as well. 

Only then will the binaries run on the target without silly errors like 'symbol lookup error, symbol not found'.    
The impact was exacerbated by the fact that an update (for rustup) broke an already built (and loaded) target system, and when rebuilt (because of an automated yocto build) didn't have 'hooks' to update the  libraries that need to be copied to the target as well.
